### PR TITLE
[Merged by Bors] - feat(Tactic/Linter/Lint): improve `dupNamespace` linter

### PR DESCRIPTION
**Pull request by mo271**
Refactor the the way the ids are extracted for the `dupNamespace` linter. We also extract this to a new file in anticipation of re-using that function in another linter.

The new version finds auto-generated names, such as `to_additive` names, but ignores "internal" names.  Like the previous linter, it also supports `export` and `alias`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
I think the import of `import Lean.DeclarationRange` causes the unpleaseant "large-import" label

**Comment by adomani** [comment]
Technically, I think that the `large-import` is a consequence of adding `Common` to `Lint`.  Still, in this case it is not such a big deal: the actual imports of `Lint` are not increased, as the extra imported file is simply to add one extra function.

**Comment by adomani** [comment]
The linter is now catching some hygienic names that have duplicated namespace since their paths have duplicated names!  This is a great test for the linter and maybe suggests that the linter should ignore hygienic names in its checks.

**Comment by adomani** [comment]
I pushed a small change to the linter and a fix for the `to_additive` test.  I'm not sure if this version will ignore those "internal" declarations that were triggering the linter, but we will see.

There is still the issue of `export`, but `alias` seems to behave as expected now.

**Comment by adomani** [comment]
The way in which I would fix the `export` issue checking, would be by extracting the information of the new declaration added in the environment extension `Lean.aliasExtension`, that `export` seems to update.  I won't have time to get around to this for several hours, so if someone else wants to give it a shot, feel free to do so!  It would be similar to how the `declRange` extension is accessed, but then it might be a little trickier to figure out what has been added, since the information is not recorded with the position where it was added.

**Comment by adomani** [comment]
Oh, those internal declarations still show up: maybe we simply have to exclude them explicitly.

File: Mathlib/Tactic/Linter/Lint.lean [hunk 1]
```diff
@@ -90,16 +92,23 @@ def getIds : Syntax → Array Syntax
...
+    for id in (← getNamesFrom (stx.getPos?.getD default)) ++ aliases do
+      let declName := id.getId
+      if declName.hasMacroScopes then continue
+      let nm := declName.components
+      let some (dup, _) := nm.zip (nm.tailD []) |>.find? fun (x, y) ↦ x == y
+        | return
+      Linter.logLint linter.dupNamespace id
+        m!"The namespace '{dup}' is duplicated in the declaration '{declName}'"
+
+#check resolveNameUsingNamespaces
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Lint.lean:111]:
```suggestion
```

File: Mathlib/Tactic/Linter/Lint.lean [hunk 2]
```diff
@@ -4,6 +4,8 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Floris van Doorn
 -/
 import Batteries.Tactic.Lint
+import Mathlib.Tactic.Linter.Common
+import Lean.Elab.Open
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Lint.lean:8]:
```suggestion
```

**Comment by mo271** [comment]
Looks good to me now!

**Comment by adomani** [comment]
!bench

**Comment by adomani** [comment]
I'm now wondering if we want to move the extraction of the aliases to the separate file as well, since it might be independently useful.

**Comment by adomani** [comment]
!bench

**Comment by grunweg** [comment]
Thanks a lot for your work in this PR - it seems the new code is much more robust w.r.t. to certain tricky edge cases!
Looking at the PR, I only see a test (changed) for the to_additive change --- but not for any of the later gotchas, such as
- internal names
- aliasses in general (right?)
- exports
Can you add tests for that - so a future person (such as you or me in five years, when we'll have forgotten all the details here) will not "simplify" the implementation again? (I'm not sure if additional comments in the implementation are useful; up to you.)

**Comment by adomani** [comment]
Dar Michael,

thank you for the review.  I think that `alias`es and `export`s were already tested before and are tested now: what other tests did you have in mind?

As for internal names, I have not been able to generate one: if you want to give it a shot, feel free to do so!

**Comment by mo271** [comment]
From my point of view this looks good to go now, and could be used in https://github.com/leanprover-community/mathlib4/pull/17580, which was the starting point for this refactor, @grunweg.

**Comment by grunweg** [comment]
I cannot find tests for `alias`es or `export`s (at least, not in `test/Lint.lean`. If you are really really sure they exist, that will suffice for me. Can you update the PR description - at least to mention `to_additive`, please? Then, I'll maintainer merge this. Thanks!

**Comment by adomani** [comment]
> I cannot find tests for `alias`es or `export`s (at least, not in `test/Lint.lean`. If you are really really sure they exist, that will suffice for me. Can you update the PR description - at least to mention `to_additive`, please? Then, I'll maintainer merge this. Thanks!

Dear Michael,

the tests that are here are:
* [export](https://github.com/leanprover-community/mathlib4/pull/17631/files#diff-a24fc4e713a823e86ff05256f10aa5ed0447d2fc2d28f1efc783fd36730be08fR82);
* [alias](https://github.com/leanprover-community/mathlib4/pull/17631/files#diff-a24fc4e713a823e86ff05256f10aa5ed0447d2fc2d28f1efc783fd36730be08fR66);
* [internal names](https://github.com/leanprover-community/mathlib4/pull/17631/files#diff-a24fc4e713a823e86ff05256f10aa5ed0447d2fc2d28f1efc783fd36730be08fR10-R25) are ignored;
* [to_additive](https://github.com/leanprover-community/mathlib4/pull/17631/files#diff-a24fc4e713a823e86ff05256f10aa5ed0447d2fc2d28f1efc783fd36730be08fR51) is *not* ignored.

I also updated the PR description.

Let me know if you think that anything else is missing!

**Comment by grunweg** [comment]
Thanks, I had overlooked the tests for `alias` and `export`. This looks good to me - thanks for all your careful effort in refining and polishing this linter!
maintainer merge

**Comment by kim-em** [comment]
bors merge
